dds/DCPS/ValueWriter.h

```diff
@@ -110,6 +110,8 @@ class OpenDDS_Dcps_Export ValueWriter {
     write_enum(name, static_cast<ACE_CDR::Long>(value));
   }
 
+  virtual void write_absent_value() = 0;
```

Comment:
I would like to understand more about the intent of this method.  From the code, it seems like it is called in an error condition.  However, this may have been how you implemented optional.

If it is just for errors, then I think the method can be removed and the resulting stream would just be in a bad state.  It may be useful to add return codes to the `vwrite` and `ValueWriter` family.

If it is for optional, I would again favor not having this method as the member can just be omitted.

Reply:
ValueWriter having awareness of an "empty optional" may be useful for an SomeOtherValueWriter even if JsonValueWriter doesn't use it.

Reply:
It is for optional member that is missing. For `JsonValueWriter` and `PrinterValueWriter`, I think it would be good to have a missing optional member noted as absent in the resulting stream, e.g. something like `OptionalField: null`. This way we know that the stream has entries for all members and don't have to check with the type definition to see what members are missing.

The side effect of this is that the `vwrite` function for dynamic data (and the generated ones for IDL types when we support `@optional` in IDL) can keep the same form as it is now. That is,
```
vw.begin_struct();
// Iterate over all members
for (...) {
  vw.begin_struct_member(...);
  vw.end_struct_member();
}
vw.end_struct();
```
Otherwise, I think for each member it'll need to check if the member doesn't have a value, and skip the calls to `begin_struct_member` and `end_struct_member` if so.

Reply:
From discussion with @sonndinh and @mitza-oci.

An XCDR2 ValueWriter with a final or appendable struct must write a boolean indicating the status of an optional member.  In order for this to work, `begin_struct_member` and `end_struct_member` must be called.

Two options exist for conveying the presence of the optional member:
1. Pass a flag to `begin_struct_member`.
2. Call a method like `write_absent_value`.

Option 2 was selected as it restrains the scope of `begin_struct_member` to dealing with the type of the member instead of the status/value of the member.

Reply:
I checked the DDS-JSON spec and it says "Unset optional members shall be omitted from the sample representation." So I think if we conform with the spec, `begin_struct_member` will need to know the status of the member so that it doesn't print the member name.

Reply:
I still see some benefits of printing something like `OptionalField: null`, for example, for human inspection. But I also see the benefits of omitting absent values like the DDS-JSON spec does to reduce the sample size when sending over the wire.